File: src/dynamicBody/character/enemy/move/StraightMove.java
package dynamicBody.character.enemy.move;

import dynamicBody.character.Character;
import dynamicBody.character.enemy.move.check.CheckMonster;
import dynamicBody.character.enemy.move.check.CheckMonsterImpl;
import dynamicBody.move.Direction;
import worldModel.RoomModel;
import worldModel.utilities.Pair;

/**
 * Class that implement MovePosMonster use when enemy's movement is Straight
 */
public class StraightMove implements MovePosMonster {

	private CheckMonster check;
	private Direction nextDir = null;

	private RoomModel currentRoom;

	/**
	 * Default constructor
	 * 
	 * @param room,      room where character is
	 * @param character, the character who need to move
	 */
	public StraightMove(RoomModel room, Character character) {
		currentRoom = room;
		check = new CheckMonsterImpl(character);
	}

	@Override
	public Pair<Integer, Integer> nextPos(Pair<Integer, Integer> pos, int speed, Direction dir) {

		Pair<Integer, Integer> nextPos = new Pair<Integer, Integer>(pos.getX() + (dir.getAbscissa() * speed),
				pos.getY() + (dir.getOrdinate() * speed));
		nextDir = dir;

		if (check.possiblePos(currentRoom, nextPos)) {
			return nextPos;
		} else {
			this.nextDir = check.changeDir(currentRoom, nextPos, dir);
			return pos;

		}

	}

	@Override
	public Direction getDirection() {
		if (nextDir == null) {
			throw new IllegalStateException(" Direction isn't Initialized ");
		}
		return nextDir;
	}
}

File: Ragazzini_Marco/CSharp_Task/DynamicBody/Character/Enemy/Move/StraightMove.cs
using System;
using System.Collections.Generic;
using System.Text;
using Model.DynamicBody.Character.Enemy.Move.Check;
using Model.DynamicBody.Move;

namespace Model.DynamicBody.Character.Enemy.Move
{
	public class StraightMove : IEnemyMovement
    {
		private ICheckEnemy Check;
		private Direction.EnumDirection nextDir;

		/// <summary>
		/// Default constructor
		/// </summary>
		/// <param name="character">the character who need to move</param>
		public StraightMove(ICharacter character)
		{
			Check = new CheckEnemyImpl(character);
		}

		public ValueTuple<int, int> NextPos(ValueTuple<int, int> pos, int speed, Direction.EnumDirection dir)
		{

			ValueTuple<int, int> nextPos = new ValueTuple<int, int>(pos.Item1 + (Direction.GetAbscissa(dir) * speed),
					pos.Item2 + (Direction.GetOrdinate(dir) * speed));
			nextDir = dir;

			if (Check.PossiblePos(nextPos))
			{
				return nextPos;
			}
			else
			{ 
				this.nextDir = Check.ChangeDir(nextPos, dir);
				return pos;

			}

		}

		public Direction.EnumDirection GetDirection()
		{
			return nextDir;
		}

    }
}
